model/jpa/src/main/java/org/keycloak/authorization/jpa/store/JPAPolicyStore.java

```diff
@@ -285,6 +287,80 @@ public void findByScopes(ResourceServer resourceServer, Resource resource, List<
                 .forEach(consumer::accept);
     }
 
+    @Override
+    public void findResourcePermissionPolicies(ResourceServer resourceServer, Resource resource, Collection<Scope> scopes, String resourceType, boolean resourceServerPolicies, Consumer<Policy> consumer) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<PolicyEntity> policyQuery = cb.createQuery(PolicyEntity.class);
+        Root<PolicyEntity> root = policyQuery.from(PolicyEntity.class);
+        policyQuery.select(root);
+
+        List<Predicate> predicates = new LinkedList<>();
+
+        Join<PolicyEntity, ResourceEntity> resourceJoin = root.join("resources", JoinType.LEFT);
+        Join<PolicyEntity, ScopeEntity> scopeJoin = root.join("scopes", JoinType.LEFT);
+        MapJoin<PolicyEntity, String, String> policyConfigJoin = root.joinMap("config", JoinType.LEFT);
+
+        // selection on resource id provided. Will also naturally overlap with scopes and types needed
+        if(resource != null) {
+            predicates.add(cb.equal(resourceJoin.get("id"), resource.getId()));
+        }
+
+        // selection on scope ids with no resource
+        if (CollectionUtil.isNotEmpty(scopes)) {
+            Predicate baseScopePredicate = scopeJoin.get("id").in(scopes.stream().map(Scope::getId).collect(Collectors.toList()));
+            baseScopePredicate = cb.and(
+                    baseScopePredicate,
+                    cb.isEmpty(root.get("resources")),
+                    // the named query excludes any policy that has a resourceType, not sure that this is expected behavior
+                    // would expect any policy with the correct scope to return regardless of type on this query
+                    cb.or(
+                            cb.isEmpty(root.get("config")),
+                            cb.notEqual(policyConfigJoin.key(), "defaultResourceType")
+                    )
+            );
+            predicates.add(baseScopePredicate);
+        }
+
+        if (StringUtil.isNotBlank(resourceType)) {
+            List<Predicate> typePredicates = new LinkedList<>();
+
+            // when this feature is enabled we should also get the policies from other resources on the server that match the resource type
+            if(resourceServerPolicies) {
+                // only applies to resources on the same resource server that are owned by the resource server and match the type
+                Predicate basePredicate = cb.and(
+                        cb.like(resourceJoin.get("type"), resourceType),
+                        cb.equal(resourceJoin.get("resourceServer"), resourceServer.getId()),
+                        cb.equal(resourceJoin.get("owner"), resourceServer.getClientId())
+                );
+
+                // if we have a resource then don't duplicate those results, only get policies from other resources
+                if(resource != null) {
```

Comment:
Just occurred to me I can remove this. Since we're working at the DB layer it's just an extra match on a record but it won't duplicate the result. Adding the extra parameter just makes processing take more time. 

This doesn't apply for other queries like scopes though. Resources have to be empty there regardless of whether or not we have a resource on the current query or not. If we want to match scopes generically then there are methods for that in this class

Reply:
on second thought this would return policies that overlap with the resource match and would duplicate cache entries resulting in duplicated runs